### PR TITLE
implement ViewModel for PackagePath dialog

### PR DESCRIPTION
**Pull request by nguyen-binh-minh**
@Benglin PTAL

### CODE REVIEW

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            MovePathUpCommand = new DelegateCommand(p => SwapPath(SelectedIndex, SelectedIndex - 1));
+            MovePathDownCommand = new DelegateCommand(p => SwapPath(SelectedIndex, SelectedIndex + 1));
+            UpdatePathCommand = new DelegateCommand(p => UpdatePathAt(p as string, SelectedIndex));
+            SaveSettingCommand = new DelegateCommand(CommitChanges, CanCommitChanges);
+        }
+
+        private bool CanDelete(object param)
+        {
+            return RootLocations.Count > 0;
+        }
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs]: Should have been `RootLocations.Count > 1`, we need at least one.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+using Dynamo.Interfaces;
+using DelegateCommand = Dynamo.UI.Commands.DelegateCommand;
+
+
+namespace Dynamo.ViewModels
+{
+    class PackagePathViewModel : ViewModelBase
+    {
+        public ObservableCollection<string> RootLocations { get; private set; }
+        public int SelectedIndex { get; set; }
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs]: In the setter of `SelectedIndex`, there should be multiple calls to `RaiseCanExecuteChanged` on the relevant command. For example, when the `SelectedIndex` is on the first or last item, move up/down will be disabled accordingly (can't move the selection item higher/lower).

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+        private void SwapPath(int x, int y)
+        {
+            if (x < 0 || y < 0 || x >= RootLocations.Count || y >= RootLocations.Count)
+                return;
+
+            var tempPath = RootLocations[x];
+            RootLocations[x] = RootLocations[y];
+            RootLocations[y] = tempPath;
+        }
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs]: This method `SwapPath` needs to update `SelectedIndex` internally to keep track on the right index.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        private void InsertPathAt(string path, int index)
+        {
+            RootLocations.Insert(index, path);
+        }
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs]: If `SelectedIndex` is larger or equal to `index`, then inserting a new item will change the selection, do handle this here (and also the `RemovePathAt` method below).

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+        private void CommitChanges(object param)
+        {
+            setting.CustomPackageFolders = new List<string>(RootLocations);
+        }
+
+        private bool CanCommitChanges(object param)
+        {
+            return RootLocations.Count > 0;
+        }
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs]: Don't quite get the idea behind this?

Comment (nguyen-binh-minh) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs]: @Benglin the initial idea was to allow users to delete everything in the list of paths but only allow them to save if there are at least 1 item in the list.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            get { return selectedIndex; }
+            set 
+            {
+                selectedIndex = value;
+                if (selectedIndex < RootLocations.Count - 1)
+                    MovePathDownCommand.RaiseCanExecuteChanged();
+                if (selectedIndex > 0)
+                    MovePathUpCommand.RaiseCanExecuteChanged();
+            }
+        }
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs]: Forget about the `if` conditions, just raise the events regardless. Also, the first curly bracket should be on a new line.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        {
+            RootLocations = new ObservableCollection<string>(setting.CustomPackageFolders);
+            SelectedIndex = 0;
+            this.setting = setting;
+
+            AddPathCommand = new DelegateCommand(p => InsertPathAt(p as string, RootLocations.Count));
+            DeletePathCommand = new DelegateCommand(p => RemovePathAt(SelectedIndex), CanDelete);
+            MovePathUpCommand = new DelegateCommand(p => SwapPath(SelectedIndex, SelectedIndex - 1), CanMoveUp);
+            MovePathDownCommand = new DelegateCommand(p => SwapPath(SelectedIndex, SelectedIndex + 1), CanMoveDown);
+            UpdatePathCommand = new DelegateCommand(p => UpdatePathAt(p as string, SelectedIndex));
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs]: The above five commands must be updated to have this signature:

```csharp
new DelegateCommand(p => SwapPath(((int) p), ((int) p) + 1), CanMoveDown);
```

Which means each `DelegateCommand` has an input as `object` but is actually an `int` (the selected item's index).

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+            }
+            set
+            {
+                selectedIndex = value;
+                MovePathDownCommand.RaiseCanExecuteChanged();
+                MovePathUpCommand.RaiseCanExecuteChanged();
+            }
+        }
+
+        public event EventHandler RequestShowFileDialog;
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs]: Try to use a new  `RequestShowFileDialogHandler` type.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        public DelegateCommand DeletePathCommand { get; private set; }
+        public DelegateCommand MovePathUpCommand { get; private set; }
+        public DelegateCommand MovePathDownCommand { get; private set; }
+        public DelegateCommand UpdatePathCommand { get; private set; }
+        public DelegateCommand SaveSettingCommand { get; private set; }
+
+        public PackagePathViewModel(IPreferences setting)
+        {
+            RootLocations = new ObservableCollection<string>(setting.CustomPackageFolders);
+            SelectedIndex = 0;
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs]: Move this to the end so all commands are non-null.

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        private int selectedIndex;
+        public int SelectedIndex
+        {
+            get
+            {
+                return selectedIndex;
+            }
+            set
+            {
+                selectedIndex = value;
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs]: `RaisePropertyChanged("SelectedIndex");`

File: src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+                return;
+
+            var tempPath = RootLocations[x];
+            RootLocations[x] = RootLocations[y];
+            RootLocations[y] = tempPath;
+
+            SelectedIndex = y;
+        }
+
+        private void InsertPathAt(int index)
```
Comment (Benglin) [inline src/DynamoCoreWpf/ViewModels/PackageManager/PackagePathViewModel.cs]: Let's insert to the last always.

**Comment by nguyen-binh-minh** [comment]
updated

**Pull request closed without merging**